Improve init.py to have a more organised structure and functionality #27 Resolved

asthacs opened the pull request:
Fixes #27 

- Files changed - `chatbot.py`, `init.py`
- Added chatbot response for an empty input by the user.
- The chatbot now exits cleanly when the user inputs "bye".

Please review @vishakha-lall

asthacs committed:
Merge pull request #2 from vishakha-lall/gssoc-master

Pull updated code

asthacs committed:
Merge pull request #3 from vishakha-lall/gssoc-master

Pulled updated code

asthacs committed:
Added response for empty input by the user

asthacs committed:
Implemented clean exit for conversation

asthacs committed:
Added response for empty input by the user

asthacs committed:
Implemented clean exit for end of conversation

vishakha-lall commented:
I see some duplicate commits. Can you check those?

asthacs commented:
I had to commit again because of some spacing issues with my code editor.
I tried to revert the old commits, but a merge conflict is coming up. Can you help me out please?
![image](https://user-images.githubusercontent.com/41833773/76792057-2b9cee80-67e8-11ea-8858-c66bc6458a67.png)

vishakha-lall commented on init.py:
```diff
@@ -1,7 +1,9 @@
+	received_message = input("You: ")
+	send_message, learn_response = message_to_bot(received_message,clf,learn_response)
+	print("MapBot: "+send_message)
+	if received_message.lower() == "bye" or received_message.lower() == "bye." or received_message.lower() == "bye!":															#exit loop
```
`chatbot.py` already handles 'bye'. Use the same to avoid redundant code here. Cleanly exiting is not implemented there, which could be added in `init.py`.

asthacs committed:
Fixed the indentation for empty input

asthacs committed:
Improved code to handle exit of conversation

asthacs commented:
I have made the required changes. I also realised there was an indentation error in the code I added in `chatbot.py`. I have fixed it now.
Please review it @vishakha-lall.

athityakumar commented on init.py:
```diff
@@ -1,7 +1,10 @@
     send_message, learn_response = message_to_bot(received_message, clf, learn_response)
     print("MapBot:  " + send_message)
+    if send_message == exit_conversation:
+    	break
```
As a guideline, leave an empty line at the end of the file 😄

athityakumar commented on chatbot.py:
```diff
@@ -35,9 +35,12 @@ def message_to_bot(H, clf, learn_response):
         learn_response = 0
         return B, learn_response
-    if H.lower() == "bye" or H.lower() == "bye." or H.lower() == "bye!":   # empty input
+    if H.lower() == "bye" or H.lower() == "bye." or H.lower() == "bye!":   
```
Hmm, would something like this be better?

```python3
# option 1
BYE_MESSAGES = ["bye", "bye.", "bye!"]
if H.lower() in BYE_MESSAGES:
    ...

# option 2
if H.lower().startswith("bye"):
    ...
```

asthacs replied:
@athityakumar how about the following?
```
if "bye" in H.lower():   
    ...
```
This would also handle cases when the user types "Okay bye" or something like that. How does that sound?

athityakumar replied:
Hmm, but that'll also have edge cases (though I don't think it's very likely that someone will use it): words like lobbyer will also get considered as a bye phrase.

athityakumar replied:
@vishakha-lall - Thoughts?

asthacs replied:
Yes, that's true. So should I go for the option 1 you suggested above then?

athityakumar replied:
Let's wait for @vishakha-lall to tell her opinion as well. I think we can go with option 2 like

```python3
if "bye" in H.lower().split(" "): # check in words within H
    ...
```

vishakha-lall replied:
Hmm, I agree with @athityakumar's latest suggestion. That should be a good implementation.

asthacs replied:
I tried to test the code using the latest suggestion, but it fails to recognize the cases where "bye!" or "bye." is used. I tried a few more different implementations, but none of them seems to work well. Could you help me out?

janakrajchadha replied:
@asthacs I think this issue is coming up since we have no text pre-processing in place which would filter out elements like punctuations, stop words etc. from the input text. That would solve the scenarios that you're facing here. Also, @athityakumar's suggestion would do a fair job in most cases but since it can run into edge cases, I'd suggest we look for some more approaches before we close this.

janakrajchadha replied:
Ideally, sentence/word similarities would be the way to go about this issue. However, that is far beyond the scope of this issue but we should add it to our 'backlog' 😛  @athityakumar @vishakha-lall

asthacs replied:
@janakrajchadha so what changes should I make on my end?

vishakha-lall replied:
You can go with the latest suggestion by @athityakumar

vishakha-lall replied:
@janakrajchadha can you create a new issue around your suggestion - to implement word/sentence similarities for this case.

athityakumar commented on init.py:
```diff
@@ -1,7 +1,10 @@
 from chatbot import setup
 from chatbot import message_to_bot
 clf, learn_response = setup()
+exit_conversation = "Bye! I'll miss you!"
```
Yet another nitpicking. Have constant values in caps.

```python3
EXIT_CONVERSATION = "Bye! I'll miss you!"
```

asthacs committed:
Fixed the representation for constants

asthacs committed:
Changed the code for handling 'bye'

asthacs commented:
I have made the changes. Please review it @vishakha-lall @athityakumar

athityakumar reviewed (approved):
LGTM! 👍

vishakha-lall reviewed (approved):
Changes look good! Well done :)
